### DISCUSSION TITLE
Naming conventions of attributes for null-state static analysis

### DISCUSSION THREAD

**Discussion by deep-outcome** [dotnet/csharplang#4915] [Language Ideas] [answered]
Traversing diagnostic `attributes` regard null-state static analysis as showed in overview at [Attributes for null-state static analysis](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/attributes/nullable-analysis#postconditions-maybenull-and-notnull), one can hit that some of them uses _if_ and some _when_ to denote its conditional implications. Quickly it is expected this has some conventional meaning but unfortunately it does not.

Attribute | Works against | Adverb
--|--|--
[`MaybeNullWhen`](https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.codeanalysis.maybenullwhenattribute?view=net-5.0) | `bool` | _When_
[`NotNullWhen`](https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.codeanalysis.notnullwhenattribute?view=net-5.0)| `bool` | _When_
[`MemberNotNullWhen`](https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.codeanalysis.membernotnullwhenattribute?view=net-5.0)|	`bool` | _When_
[`NotNullIfNotNull`](https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.codeanalysis.notnullifnotnullattribute?view=net-5.0)| `null` | _If_ (!)	
[`DoesNotReturnIf`](https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.codeanalysis.doesnotreturnifattribute?view=net-5.0)| `bool` | _If_ (!)

So, I guess this is some kind of microchaos. Why not all of them uses _When_/_If_? Why if both of them is used, there is no order about choose of one?

**Comment by yaakov-h** [comment, marked as answer]
“When” relates to the result of method execution. “If” relates to an input value.

**Reply by deep-outcome** [reply]
Yes, yes. It makes sense.

**Comment by ZacharyPatten** [comment]
The names make sense to me, but if you don't like the names, you can always use an alias:
```cs
using dog = System.Diagnostics.CodeAnalysis.MaybeNullAttribute;
using cat = System.Diagnostics.CodeAnalysis.NotNullAttribute;
using bird = System.Diagnostics.CodeAnalysis.NotNullWhenAttribute;

public class C
{
	[return: dog]
	public T Find<T>() => default(T);

	public static void ThrowWhenNull([cat] object? value) { }

	void IsNullOrEmpty([bird(false)] string? value) { }
}
```
I wouldn't recommend doing that though, as it would make your code harder for others to read.
